src/MILTools.hs

```diff
@@ -58,16 +57,20 @@ milCompile milo outputFileName invokeClang (prog, entrypoints) =
 
        if fake milo
        then putStrLn milCmd
-       else do exitCode <- system milCmd
-               if exitCode /= ExitSuccess
-               then hPutStrLn stderr ("mil-tools invocation failed (" ++ show exitCode ++ ")")
-               else do removeFile lcFileName
-                       if invokeClang
-                       then do exitCode <- system clangCmd
-                               if exitCode /= ExitSuccess
-                               then hPutStrLn stderr ("clang invocation failed (" ++ show exitCode ++")")
-                               else removeFile llFileName
-                       else return ()
+       else runExternal milCmd
+            (do removeFile lcFileName
+                runExternal clangCmd
+                  (removeFile llFileName))
 
     where lcFileName = replaceExtension outputFileName "lc"
           llFileName = replaceExtension outputFileName "ll"
+
+runExternal :: String -> IO () -> IO ()
+runExternal cmd next =
+  do exitCode <- system cmd
+     case exitCode of
+       ExitSuccess -> next
+       _           -> showError cmd exitCode
+
+showError :: String -> ExitCode -> IO ()
```

Comment:
this can be an inner function if it's being used exclusively used only by `runExternal`
looks good to me otherwise..